content/post/scaling-workloads/index.md

```diff
@@ -0,0 +1,116 @@
+On many modern systems, the CPU has cache and the processor will read from that. Fetching also results in
+the CPU incrementing the program counter. This is in preparation for the next time a fetch occurs. The
+CPU then has to decode the instruction which results in the control unit setting various timing and control
+signals. Many modern CPU's use microcode which the microcode results in controlling the control unit to
+set the timing pulses and control signals. Once the instruction is decoded, the CPU must execute it.
+The execution step results in various signals being sent depending on the instruction and various control
+signals being set. Now, let's look at just how this can result in a simple program in x86 assembly can
+result in suboptimal performance.
+
+```asm
+mov ax, 1
+mov bx, 1
+add ax, bx
+```
+
+This simple program consists of a total of 3 instructions. Two instructions are a move (`mov`) and a single
+addition (`add`). For this example, we are dealing with 8-bit numbers which is the smallest bit size x86
+supports. This program results in 3 full clock cycles to occur. The first two set registers `ax` and `bx` to `1`.
+The second then performs an addition. This program would look like this in a math setting.
+
+```js
+ax = 1
+bx = 1
+ax = ax + bx
+```
+
+Once we get to the last instruction which is an addition instruction, this results in the ALU adding `ax`
+and `bx`. The value from the ALU is then put back into `ax`. If instead of using `ax`, we were to use
+a memory address then this will result in a higher execution time. Memory is always going to be slower
+than accessing a register. This is due to the registers being gates inside the silicon and therefore
+the lines from the 8-bit:8-bit ALU (the ALU optimized for performing an operation on 2 8-bit numbers)
+to the CPU being significantly shorter. Just from the looks of it, this program may not be that complex
+like a compiler but there is always a chance this program will not run optimally. In fact, this program
+isn't as optimal as it can be for an x86 program. The most optimal is simply this.
+
+```asm
+mov ax, 1
+add ax, 1
+```
+
+We can further optimize this if we always assume `ax` will be `1` before we run the `add` instruction.
+This results in a program which looks like this.
+
+```asm
+add ax, 1
+```
+
+This program results in a single clock cycle and takes the least amount of execution time compared
+to the other variants of this program. However, the optimization of removing the `mov` to `ax` can
+be dangerous. If `ax` is not `1` then it would result in a different result being stored into `ax`.
+This is an example of an optimization which can be dangerous. It is considered dangerous as it
+is assuming the CPU state being a certain way with no checks. Adding checks can result in degrated
+performance. This is why many program which deals with high performance may use hand written assembly.
```

Comment:
This sentence will need reworked, it *will* degrade performance, but it would only be really measurable on CPUs older than both of us. See also the anecdote about a 4 GHz processor will complete 4 clock cycles in the time it takes for the light from your screen to reach your eyes. 

The sentence after is a non sequitur, `This is why many program which deals with high performance` is unrelated to adding checks to prevent bad assumptions when writing assembly.